### PR TITLE
feat: update replica selection logic to use node affinity rules, toleration, node capacity

### LINKED ISSUE #230: bug: node selection logic needs to respect each node's remaining attachable capacity, volumes' node affinity rules and node taints/tolerations
**Is your feature request related to a problem?/Why is this needed**

* Current node selection logic only partially takes node capacity into account by counting the number of `AzVolumeAttachment` per node but it fails to capture nodes' remaining capacity. That said, even after node reaches its max attachable volumes limit, the controller will unknowingly schedule a replica attachment on that node.
* Volumes' node affinity rules need to be respected when selecting nodes for replica attachments of these volumes.
*  Node taints and pod tolerations need to be respected when selecting nodes for replica attachments.
 
**Describe the solution you'd like in detail**

* Controller add nodes' status / labels and volumes' node affinity rules into the equation of getting a ranked list of nodes for replica attachment.
 
**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by sunpa93**
**What type of PR is this?**
/kind update

**What this PR does / why we need it**:
* current node selection logic only partially takes node capacity into account and does not respect the number of remaining attachable disk for each node when looking for nodes for replica attachments.
* and it also does not take node affinity rules and pod tolerations into account.
* This update allows controllers to use these information to come up the candidate node list for replica attachments.
* This version of PR also addresses compatibility issue with pre-provisioning volume cases, where the PV's name and AzVolume's name do not match

**Which issue(s) this PR fixes**:

Fixes #230

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: pkg/apis/azuredisk/v1alpha1/types.go [hunk 1]
```diff
@@ -54,6 +54,9 @@ const (
 
 // AzVolumeSpec is the spec for an AzVolume resource
 type AzVolumeSpec struct {
+	// the name of PV that corresponds to the AzVolume obj
+	//+optional
+	PersistentVolume string `json:"persistentVolume"`
```
Comment (edreed) [inline pkg/apis/azuredisk/v1alpha1/types.go:59]: Since this isn't set by the `CrdProvisioner` but rather by the `pv` controller as part of its reconciliation cycle, I believe this should be part of the `Status` field rather than the `Spec` field.

File: test/e2e/testsuites/dynamically_provisioned_pod_node_selector.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	"k8s.io/kubernetes/test/e2e/framework"
+	azDiskClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	consts "sigs.k8s.io/azuredisk-csi-driver/pkg/azureconstants"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/azureutils"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/controller"
+	"sigs.k8s.io/azuredisk-csi-driver/test/e2e/driver"
+)
+
+//  will provision required PV(s), PVC(s) and Pod(s)
+// Pod should successfully be re-scheduled on failover in a cluster with AzDriverNode and AzVolumeAttachment resources
```
Comment (edreed) [inline test/e2e/testsuites/dynamically_provisioned_pod_node_selector.go:41]: Comment doesn't seem to apply to this test.

File: test/e2e/testsuites/dynamically_provisioned_pod_node_selector.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		framework.ExpectNoError(err)
+		framework.ExpectNotEqual(pv.Spec.CSI, nil)
+		framework.ExpectEqual(pv.Spec.CSI.Driver, consts.DefaultDriverName)
+
+		diskName, err := azureutils.GetDiskName(pv.Spec.CSI.VolumeHandle)
+		framework.ExpectNoError(err)
+		diskNames[i] = strings.ToLower(diskName)
+	}
+
+	// confirm that primary and replica AzVolumeAttachment nodes for this volume is not created on a node without the taint
```
Comment (edreed) [inline test/e2e/testsuites/dynamically_provisioned_pod_node_selector.go:101]: np: double-negative makes this hard to parse and doesn't seem to be quite correct. Maybe "Confirm that the primary and replica AzVolumeAttachments for this volume are created on a node without the label"?

File: test/e2e/testsuites/dynamically_provisioned_pod_toleration.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	// Get the list of available nodes for scheduling the pod
+	nodes := ListNodeNames(client)
+	if len(nodes) < maxMountReplicaCount+1 {
+		ginkgo.Skip("need at least %d nodes to verify the test case. Current node count is %d", maxMountReplicaCount+1, len(nodes))
+	}
+
+	ctx := context.Background()
+
+	// set node taint
+	numNodesWithTaint := len(nodes) - (maxMountReplicaCount + 1)
```
Comment (edreed) [inline test/e2e/testsuites/dynamically_provisioned_pod_toleration.go:62]: We general have 3 node clusters for e2e tests: 1 master and 2 workers. The master node already carries a taint which I believe would prevent pods from being scheduled. If we label or taint one of the workers, we won't have enough nodes for replicas. Should we increase the work node counts to make more available?

File: test/e2e/testsuites/dynamically_provisioned_pod_toleration.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	nodesWithTaint := map[string]struct{}{}
+	for i := 0; i < numNodesWithTaint; i++ {
+		nodeObj, err := client.CoreV1().Nodes().Get(ctx, nodes[i], metav1.GetOptions{})
```
Comment (edreed) [inline test/e2e/testsuites/dynamically_provisioned_pod_toleration.go:65]: We're labelling or tainting nodes in the order returned by `ListNodeNames` which includes the master node. Given the master node is already tainted to prevent schedule workloads (by default), `ListNodeNames` should filter out nodes with the `node-role.kubernetes.io/master` label (or only include nodes with the `node-role.kubernetes.io/agent` label), and probably be renamed to `ListAgentNodes`.

File: test/e2e/testsuites/dynamically_provisioned_pod_toleration.go (hunk 4 above)
Comment (sunpa93) [reply test/e2e/testsuites/dynamically_provisioned_pod_toleration.go:62]: Or we could also add toleration to the pod to allow scheduling on master node as well.

Comment (edreed) [reply test/e2e/testsuites/dynamically_provisioned_pod_toleration.go:62]: I am fine with either approach.

File: test/e2e/dynamic_provisioning_test.go [hunk 6]
```diff
@@ -856,6 +857,108 @@ func (t *dynamicProvisioningTestSuite) defineTests(isMultiZone bool, schedulerNa
...
+		test := testsuites.PodNodeSelector{
+			CSIDriver:              testDriver,
+			Pod:                    pod,
+			AzDiskClient:           azDiskClient,
+			Volume:                 volume,
+			StorageClassParameters: storageClassParameters,
+		}
+		test.Run(cs, ns, schedulerName)
+	})
+
```
Comment (edreed) [inline test/e2e/dynamic_provisioning_test.go:961]: We should also have a test case for pod affinity/anti-affinity.

**Comment by edreed** [comment]
@sunpa93 It looks like you update the "latest" chart package instead of the v2.0.0-alpha.1 chart. This is why the Verify tests are failing. You should revert the "latest" chart package, rebuild the v2.0.0-alpha.1 chart and reindex the chart repo.

```bash
helm package charts/v2.0.0-alpha.1/azuredisk-csi-driver -d charts/v2.0.0-alpha.1/
helm repo index charts --url=https://raw.githubusercontent.com/abhisheksinghbaghel/azuredisk-csi-driver/master/charts
```

**Comment by sunpa93** [comment]
>>  It looks like you updated the "latest" chart package instead of the v2.0.0-alpha.1 chart. This is why the Verify tests are failing. You should revert the "latest" chart package, rebuild the v2.0.0-alpha.1 chart and reindex the chart repo.

@edreed The thing is I did update the v2.0.0-alpha.1 chart and the verify was still complaining, which was why I tried updating the `latest` chart package as well given that the error message was suggesting `latest chart config has changed, pls run "helm package charts/latest/azuredisk-csi-driver -d charts/latest/" to update tgz file`. Hmm

**Comment by edreed** [comment]
> @edreed The thing is I did update the v2.0.0-alpha.1 chart and the verify was still complaining, which was why I tried updating the `latest` chart package as well given that the error message was suggesting `latest chart config has changed, pls run "helm package charts/latest/azuredisk-csi-driver -d charts/latest/" to update tgz file`. Hmm

I don't see a new package for v2.0.0-alpha.1 in the change; just the one for latest.

**Comment by sunpa93** [comment]
> I don't see a new package for v2.0.0-alpha.1 in the change; just the one for latest.

(Slap on the head). Just realized that I ran `helm package` on v2.0.0-alpha.1 without the `-d` flag.

> BTW- latest != v2.0.0-alpha.1, but rather the latest version of the v1 driver.

Yes, I am aware :D

File: test/e2e/testsuites/testsuites.go [hunk 7]
```diff
@@ -949,6 +999,53 @@ func ListNodeNames(c clientset.Interface) []string {
...
+	} else {
+		// acknowledge that this can lead to duplicate taints
+		node.Spec.Taints = append(node.Spec.Taints, taints...)
+	}
+
+	_, err = c.CoreV1().Nodes().Update(context.Background(), node, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Errorf("failed to add taints (%+v) to node (%s): %v", taints, node.Name, err)
+	} else {
+		klog.Infof("successfullly added taints (%+v) to node (%s)", taints, node.Name)
```
Comment (edreed) [inline test/e2e/testsuites/testsuites.go:1044]: sp: successfully - one too many 'l's.

**Comment by edreed** [comment]
@sunpa93 Since all the tests but Verify succeeded prior to your latest commit, you only changed a spelling error in a log message and Verify is now passing, you have my approval to check in before the remaining tests complete.
